packages/node_modules/@ciscospark/internal-plugin-ediscovery/src/report-generator.js

```diff
@@ -0,0 +1,133 @@
+
+    if (!attrs || !attrs.reportId) {
+      throw Error('Missing Required Argument reportId');
+    }
+
+    this.reportId = attrs.reportId;
+  },
+
+  /**
+   * Map an activity to an MIME rfc822 formated string that can be saved as an eml file.
+   * @param {Object} activity - An activity object
+   * @returns {String} A string that represents the data stored in an eml file.
+   */
+  map(activity) {
+    if (!activity) {
+      throw Error('Invalid activity');
+    }
+
+    const mime = new MimeBuilder('multipart/mixed')
+      .setHeader('From', activity.actorDisplayName)
+      .setHeader('Subject', activity.activityId)
+      .setHeader('Date', activity.timestamp);
+
+    let content = activity.objectDisplayName;
+    if (activity.verb !== 'post') {
+      content = activity.verb;
+      // Include the target of the event. e.g. Add Person
+      if (activity.objectId) {
+        content += ` ${activity.objectId}`;
+      }
+    }
+
+    mime.createChild('text/plain')
+      .setContent(content);
+
+    if (activity.verb === 'share' && activity.files) {
+      for (let i = 0; i < activity.files.length; i += 1) {
+        const file = activity.files[i];
+        if (file.name && file.content) {
+          mime.createChild(file.mimeType, {filename: file.name})
+            .setContent(file.content)
+            .setHeader('Content-Disposition', 'attachment');
+        }
+      }
+    }
+
+    return mime.build();
+  },
+
+  /**
+   * Add activities to the report.
+   * @param {Object[]} activities - Activities that are to form part of the report
+   * @returns {Void}
+   */
+  add(activities) {
+    if (!activities || !Array.isArray(activities) || !activities.length) {
+      throw Error('No supplied activities to add');
+    }
+
+    if (!this.zip) {
+      this.zip = new JSZip();
+      // Top level zip folder is based on the report id
+      this.folder = this.zip.folder(`${this.reportId}`);
+      this.spaces = {};
+    }
+
+    for (let i = 0; i < activities.length; i += 1) {
+      const activity = activities[i];
+
+      // If there is no target id use the report id folder
+      let space = this.folder;
+      if (activity.targetId) {
+        if (!(activity.targetId in this.spaces)) {
+          // Create sub-folder for space id
+          this.spaces[activity.targetId] = this.folder.folder(`${activity.targetId}`);
+        }
+
+        space = this.spaces[activity.targetId];
+      }
+
+      const data = this.map(activity);
+      space.file(`${activity.activityId}.eml`, data);
+    }
+  },
+
+  /**
+   * Once all the activities have been added create a report and save it to disk.
+   * @returns {Promise} A promise on the generated report
+   */
+  saveAs() {
+    if (!this.zip) {
+      throw Error('No activities added to report');
+    }
+
+    return this.zip.generateAsync({type: 'blob'})
+      .then((content) => {
+        // Download to the user's browser
+        saveAs(content, `${this.reportId}.zip`);
```

Comment:
is there a reason to use this over the native `fs` methods? I'm not against bringing in dependencies, just trying to understand it better.

Reply:
I think @gannons may be in a better position to discuss that

Reply:
The native fs module is stripped when the sdk is wrapped by browserify. i.e. As the browse does not allow direct file system access. There is a replacement module called browserify-fs. But it works on an in-memory file system.

The saveAs(..) function prompts the browse to kick off a download.